File: android/src/main/java/com/margelo/nitro/rive/DataLoader.kt
package com.margelo.nitro.rive

sealed class DataLoaderException(message: String) : Exception(message) {
  object InvalidSource : DataLoaderException("Invalid data source type for this loader")
  class InvalidURL(url: String) : DataLoaderException("Invalid URL: $url")
  class HttpError(val statusCode: Int, url: String) :
    DataLoaderException("HTTP error $statusCode for $url")
  class ResourceNotFound(resource: String) : DataLoaderException("Resource not found: $resource")
  class FileNotFound(path: String) : DataLoaderException("File not found: $path")
  object NoContext : DataLoaderException("No application context available")
}

interface DataLoader {
  suspend fun load(source: DataSource): ByteArray
}

File: ios/DataLoader.swift
import Foundation

enum DataLoaderError: Error, LocalizedError {
  case invalidSource
  case invalidURL(String)
  case httpError(statusCode: Int, url: URL)
  case resourceNotFound(resource: String)
  case fileNotFound(path: String)

  var errorDescription: String? {
    switch self {
    case .invalidSource:
      return "Invalid data source type for this loader"
    case .invalidURL(let url):
      return "Invalid URL: \(url)"
    case .httpError(let code, let url):
      return "HTTP error \(code) for \(url)"
    case .resourceNotFound(let resource):
      return "Resource not found: \(resource)"
    case .fileNotFound(let path):
      return "File not found: \(path)"
    }
  }
}

protocol DataLoader {
  func load(from source: DataSource) async throws -> Data
}
